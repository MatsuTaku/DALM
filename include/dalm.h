#ifndef DALM_H_
#define DALM_H_

#include <string>
#include <fstream>
#include <vector>
#include <set>

#include "logger.h"
#include "vocabulary.h"
#include "da.h"
#include "version.h"
#include "state.h"

typedef unsigned long int StateId;

namespace DALM {
	class Fragment;

	class LM {
		friend class Fragment;

		public:
			LM(std::string pathtoarpa, std::string pathtotreefile, Vocabulary &vocab, size_t dividenum, Logger &logger);
			LM(std::string dumpfilepath, Vocabulary &vocab, Logger &logger);
			virtual ~LM();

			float query(VocabId *ngram, size_t n);
			float query(VocabId word, State &state);

			float query(VocabId word, State &state, Fragment &f);
			float query(const Fragment &f, State &state_prev, Gap &gap);
			float query(const Fragment &fprev, State &state_prev, Gap &gap, Fragment &fnew);

			void init_state(State &state);
			void set_state(VocabId *ngram, size_t n, State &state);
			void set_state(State &state, State &state_prev, Gap &gap);
			
			/* depricated */
			StateId get_state(VocabId *ngram, size_t n);

			void dump(std::string dumpfilepath){
				logger << "[LM::dump] start dumping to file(" << dumpfilepath << ")" << Logger::endi;
				FILE *fp = std::fopen(dumpfilepath.c_str(),"wb");
				Version v;
				v.dump(fp);
				value_array->dump(fp);
				dumpParams(fp);
				fclose(fp);
				logger << "[LM::dump] File-dump done." << Logger::endi;
			}

		private:
			void errorcheck(std::string &pathtoarpa);
			void build(std::string &pathtoarpa, std::string &pathtotreefile, size_t dividenum);
			void dumpParams(FILE *fp);
			void readParams(FILE *fp);

			DA **da;
			ValueArray *value_array;
			unsigned char danum;
			Vocabulary &vocab;
			Logger &logger;
	};
}

#endif /* DALM_H_ */
