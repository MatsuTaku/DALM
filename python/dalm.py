from _dalm import State, LM, Vocabulary, Model, _build

class Builder(object):
    bst = "bst"
    reverse = "reverse"

    def __init__(self, memory_limit=1024, n_cores=1):
        self.memory_limit = memory_limit
        self.n_cores = n_cores

    def __call__(self, output, arpa, opt="reverse", n_div=1):
        _build(output, arpa, opt, n_div, self.n_cores, self.memory_limit)

