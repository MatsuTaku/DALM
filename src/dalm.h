#ifndef DALM_H_
#define DALM_H_

#include <string>
#include <fstream>
#include <vector>
#include <memory>
#include <set>
#include <glob.h>
#include "dalm/vocabulary.h"
#include "dalm/logger.h"
#include "dalm/state.h"
#include "dalm/fragment.h"
#include "dalm/fileutil.h"
#include "dalm/da.h"
#include "dalm/version.h"
#include "dalm/handler.h"
#include "dalm/vocabulary_file.h"

typedef unsigned long int StateId;
namespace DALM{
  namespace QUANT_INFO{
    extern int prob_bits;
    extern int bow_bits;
    extern int independent_left_mask;
    extern int prob_mask;
    extern std::vector<float> prob_bin[DALM_MAX_ORDER];
    extern std::vector<float> bow_bin[DALM_MAX_ORDER];
  }
}

namespace DALM {
    class LM {
        public:
            LM(std::string pathtoarpa, std::string pathtotreefile, Vocabulary &vocab, unsigned int dividenum, unsigned int opt, std::string tmpdir, unsigned int n_cores, Logger &logger, int transpose, int prob_bit, int bow_bit);
            LM(std::string dumpfilepath, Vocabulary &vocab, unsigned char order, Logger &logger);
            virtual ~LM();
            
            float query(VocabId *ngram, unsigned char n);
            float query(VocabId word, State &state);

            float query(VocabId word, State &state, Fragment &f);
            float query(const Fragment &f, State &state_prev, Gap &gap);
            float query(const Fragment &fprev, State &state_prev, Gap &gap, Fragment &fnew);
            float update(VocabId word, Fragment &f, bool &extended, bool &inedependent_left, bool &independent_right, float &bow);


            void init_state(State &state);
            void set_state(VocabId *ngram, unsigned char n, State &state);
            void set_state(State &state, const State &state_prev, const Fragment *fragments, const Gap &gap);

            float sum_bows(State &state, unsigned char begin, unsigned char end);
            
            /* depricated */
            StateId get_state(VocabId *ngram, unsigned char n);
            
            void dump(std::string dumpfilepath){
                logger << "[LM::dump] start dumping to file(" << dumpfilepath << ")" << Logger::endi;
                BinaryFileWriter writer(dumpfilepath);
                v.dump(writer);
                if(DALM::QUANT_INFO::prob_bits != 0){
                    writer << DALM::QUANT_INFO::prob_bits;
                    writer << DALM::QUANT_INFO::bow_bits;
                    for(int i = 0; i < DALM_MAX_ORDER-1; ++i){
                        writer.write_many(DALM::QUANT_INFO::prob_bin[i].data(), 1<<(DALM::QUANT_INFO::prob_bits));
                        writer.write_many(DALM::QUANT_INFO::bow_bin[i].data(), (1<<(DALM::QUANT_INFO::bow_bits)));
                    }
                }
                dumpParams(writer);
                logger << "[LM::dump] File-dump done." << Logger::endi;
            }

            void export_to_text(std::string exportfilepath){
                logger << "[LM::dump] start exporting to file(" << exportfilepath << ")" << Logger::endi;
                TextFileWriter writer(exportfilepath);
                exportParams(writer);
                logger << "[LM::dump] File-dump done." << Logger::endi;

            }
            std::size_t errorcheck(std::string &pathtoarpa);

        private:
            void build(std::string &pathtoarpa, std::string &pathtotreefile, unsigned int dividenum, std::string &tmpdir, unsigned int n_cores, int transpose, int prob_bit, int bow_bit);
            void dumpParams(BinaryFileWriter &writer);
            void exportParams(TextFileWriter &writer);
            void readParams(BinaryFileReader &reader, unsigned char order);

            Vocabulary &vocab;
            Version v;
            DAHandler *handler;
            Logger &logger;
            VocabId stagid;
    };

    class Model {
    public:
        Model(std::string modeldir, unsigned char order, Logger &logger);
        Model(std::string modeldir, Logger &logger);
        virtual ~Model(){
            delete lm;
            delete vocab;
            delete vocab_file;
        }
        std::string model;
        std::string words;
        std::string wordstxt;
        Vocabulary *vocab;
        LM *lm;
        VocabularyFile *vocab_file;
        unsigned char order;
		std::string modeldir;
        Logger &logger;
    private:
        void load();
    };
}

#endif /* DALM_H_ */

