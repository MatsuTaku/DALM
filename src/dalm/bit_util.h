#ifndef BIT_UTIL_H_
#define BIT_UTIL_H_

#ifdef _MSC_VER
#include <intrin.h>
#else
#include <x86intrin.h>
#endif

namespace DALM::bit_util {

// ctz (Count trailing zeros) (***10000 = 4)
inline unsigned long long ctz64(unsigned long long x);
inline unsigned long long ctz(unsigned long long x) {
    return ctz64(x);
}
inline unsigned int ctz32(unsigned int x) {
#ifdef __BMI__
    return _tzcnt_u32(x);
#elif defined(__POPCNT__)
    return _mm_popcnt_u32((x & -x) - 1);
#else
    // Arrange of nlz algorithm in http://www.nminoru.jp/~nminoru/programming/bitcount.html#leading-0bits
    if (x == 0)
        return 32;
    unsigned int c = 0;
    if (x & 0x0000FFFF) {
        x &= 0x0000FFFF;
        c |= 16;
    }
    if (x & 0x00FF00FF) {
        x &= 0x00FF00FF;
        c |= 8;
    }
    if (x & 0x0F0F0F0F) {
        x &= 0x0F0F0F0F;
        c |= 4;
    }
    if (x & 0x33333333) {
        x &= 0x33333333;
        c |= 2;
    }
    if (x & 0x55555555) {
        c |= 1;
    }
    return c ^ 31;
#endif
}
inline unsigned long long ctz64(unsigned long long x) {
#ifdef __BMI__
    return _tzcnt_u64(x);
#elif defined(__POPCNT__)
    return _mm_popcnt_u64((x & -x) - 1);
#else
    // Arrange of nlz algorithm in http://www.nminoru.jp/~nminoru/programming/bitcount.html#leading-0bits
    if (x == 0)
        return 64;
    unsigned long long c = 0;
    if (x & 0x00000000FFFFFFFF) {
        x &= 0x00000000FFFFFFFF;
        c |= 32;
    }
    if (x & 0x0000FFFF0000FFFF) {
        x &= 0x0000FFFF0000FFFF;
        c |= 16;
    }
    if (x & 0x00FF00FF00FF00FF) {
        x &= 0x00FF00FF00FF00FF;
        c |= 8;
    }
    if (x & 0x0F0F0F0F0F0F0F0F) {
        x &= 0x0F0F0F0F0F0F0F0F;
        c |= 4;
    }
    if (x & 0x3333333333333333) {
        x &= 0x3333333333333333;
        c |= 2;
    }
    if (x & 0x5555555555555555) {
        c |= 1;
    }
    return c ^ 63;
#endif
}


// clz (Count leading zeros) (00001*** = 4)
inline unsigned long long clz64(unsigned long long x);
inline unsigned long long clz(unsigned long long x) {
    return clz64(x);
}
inline unsigned int clz32(unsigned int x) {
#ifdef __LZCNT__
    return _lzcnt_u32(x);
#elif defined(__POPCNT__)
    x |= x >> 1;
	x |= x >> 2;
	x |= x >> 4;
	x |= x >> 8;
	x |= x >> 16;
	return _mm_popcnt_u32(~x);
#else
    // Reference to http://www.nminoru.jp/~nminoru/programming/bitcount.html#leading-0bits
    if (x == 0)
        return 32;
    unsigned int c = 0;
    if (x & 0xFFFF0000) {
        x &= 0xFFFF0000;
        c |= 16;
    }
    if (x & 0xFF00FF00) {
        x &= 0xFF00FF00;
        c |= 8;
    }
    if (x & 0xF0F0F0F0) {
        x &= 0xF0F0F0F0;
        c |= 4;
    }
    if (x & 0xCCCCCCCC) {
        x &= 0xCCCCCCCC;
        c |= 2;
    }
    if (x & 0xAAAAAAAA) {
        c |= 1;
    }
    return c ^ 31;
#endif
}
inline unsigned long long clz64(unsigned long long x) {
#ifdef __LZCNT__
	return _lzcnt_u64(x);
#elif defined(__POPCNT__)
	x |= x >> 1;
	x |= x >> 2;
	x |= x >> 4;
	x |= x >> 8;
	x |= x >> 16;
	x |= x >> 32;
	return _mm_popcnt_u64(~x);
#else
	// Reference to http://www.nminoru.jp/~nminoru/programming/bitcount.html#leading-0bits
	if (x == 0)
		return 64;
	unsigned long long c = 0;
	if (x & 0xFFFFFFFF00000000) {
		x &= 0xFFFFFFFF00000000;
		c |= 32;
	}
	if (x & 0xFFFF0000FFFF0000) {
		x &= 0xFFFF0000FFFF0000;
		c |= 16;
	}
	if (x & 0xFF00FF00FF00FF00) {
		x &= 0xFF00FF00FF00FF00;
		c |= 8;
	}
	if (x & 0xF0F0F0F0F0F0F0F0) {
		x &= 0xF0F0F0F0F0F0F0F0;
		c |= 4;
	}
	if (x & 0xCCCCCCCCCCCCCCCC) {
		x &= 0xCCCCCCCCCCCCCCCC;
		c |= 2;
	}
	if (x & 0xAAAAAAAAAAAAAAAA) {
		c |= 1;
	}
	return c ^ 63;
#endif
}

}

#endif //BIT_UTIL_H_
