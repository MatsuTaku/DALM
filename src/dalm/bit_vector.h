#ifndef DALM_BIT_VECTOR_H_
#define DALM_BIT_VECTOR_H_

namespace DALM {

template <bool DefaultBit>
class BitVector {
  class _Ref {
   private:
    uint64_t *pointer_;
    size_t ctz_;
   public:
    explicit _Ref(uint64_t* pointer, size_t ctz) : pointer_(pointer), ctz_(ctz) {}
    explicit operator bool() const {
      return *pointer_ & (1ull << ctz_);
    }
    _Ref& operator=(bool bit) {
      if (bit)
        *pointer_ |= (1ull << ctz_);
      else
        *pointer_ &= compl (1ull << ctz_);
      return *this;
    }
  };

 public:
  static constexpr uint64_t kDefaultWord = DefaultBit ? -1ull : 0;

  BitVector() : size_(0) {}
  _Ref operator[](size_t index) {
    return _Ref(&container_[index / 64], index % 64);
  }
  uint64_t* ptr_at(size_t word_index) {
    return &container_[word_index];
  }
  uint64_t word_at(size_t word_index) const {
    return word_index < container_.size() ? container_[word_index] : kDefaultWord;
  }
  uint64_t bits64_from(size_t index) const {
    auto insets = index%64;
    if (insets == 0) {
      return word_at(index/64);
    } else {
      return (word_at(index/64) >> insets) | (word_at(index/64+1) << (64-insets));
    }
  };
  void resize(size_t new_size, bool bit) {
    container_.resize((new_size-1)/64+1, bit ? -1ull : 0);
    if (new_size > size_ and bit) {
      *ptr_at(size_/64) |= -1ull << size_%64;
      if (new_size % 64 > 0) {
        container_[new_size/64] &= -1ull >> (64 - new_size%64);
      }
    }
    size_ = new_size;
  }

 private:
  size_t size_;
  std::vector<uint64_t> container_;
};

}

#endif //DALM_BIT_VECTOR_H_
