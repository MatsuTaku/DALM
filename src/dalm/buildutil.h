#include<string>

namespace DALM{
    class Vocabulary;
    class Logger;
    void build_bst(std::string tempdir, std::string arpafile, std::string outpath, Vocabulary &vocab, Logger &logger, std::size_t memory_limit = 1024*1024*10);
    void build_dalm(std::string outdir, std::string arpa, std::string opt, unsigned char n_div, int n_cores, std::size_t memory_limit, Logger &logger);
}

