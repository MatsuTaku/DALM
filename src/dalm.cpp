#include <cmath>

#include <unistd.h>
#include <vector>
#include <glob.h>
#include <sys/stat.h>

#include "dalm.h"
#include "dalm/embedded_da.h"
#include "dalm/reverse_da.h"
#include "dalm/bst_da.h"
#include "dalm/pt_da.h"
#include "dalm/qpt_da.h"

namespace DALM{
  namespace QUANT_INFO{
    int prob_bits;
    int bow_bits;
    int independent_left_mask;
    int prob_mask;
    std::vector<float> prob_bin[DALM_MAX_ORDER];
    std::vector<float> bow_bin[DALM_MAX_ORDER];
  }
}

using namespace DALM;

LM::LM(std::string pathtoarpa, std::string pathtotree, Vocabulary &vocab, unsigned int dividenum, unsigned int opt, std::string tmpdir, unsigned int n_cores, Logger &logger, int transpose, int prob_bit, int bow_bit)
        :vocab(vocab), v(Version(opt, logger)), logger(logger) {
    logger << "[LM::LM] TEXTMODE begin." << Logger::endi;
    logger << "[LM::LM] pathtoarpa=" << pathtoarpa << Logger::endi;
    logger << "[LM::LM] pathtotree=" << pathtotree << Logger::endi;

    build(pathtoarpa, pathtotree, dividenum, tmpdir, n_cores, transpose, prob_bit, bow_bit);
    std::string stagstart = "<s>";
    stagid = vocab.lookup(stagstart.c_str());
    logger << "[LM::LM] end." << Logger::endi;
}

LM::LM(std::string dumpfilepath,Vocabulary &vocab, unsigned char order, Logger &logger)
    :vocab(vocab), logger(logger){

    logger << "[LM::LM] BINMODE begin." << Logger::endi;
    logger << "[LM::LM] dumpfilepath=" << dumpfilepath << Logger::endi;

    BinaryFileReader reader(dumpfilepath);
    v = Version(reader, logger); // Version check.
    readParams(reader, order);
    std::string stagstart = "<s>";
    stagid = vocab.lookup(stagstart.c_str());
    logger << "[LM::LM] end." << Logger::endi;
}

LM::~LM() {
    delete handler;
}

float LM::query(VocabId *ngram, unsigned char n){
    return handler->get_prob(ngram, n);
}

float LM::query(VocabId word, State &state){
    return handler->get_prob(word, state);
}

float LM::query(VocabId word, State &state, Fragment &f){
    return handler->get_prob(word, state, f);
}

float LM::query(const Fragment &f, State &state_prev, Gap &gap){
    return handler->get_prob(f, state_prev, gap);
}

float LM::query(const Fragment &fprev, State &state_prev, Gap &gap, Fragment &fnew){
    return handler->get_prob(fprev, state_prev, gap, fnew);
}

float LM::update(VocabId word, Fragment &f, bool &extended, bool &independent_left, bool &independent_right, float &bow) {
    return handler->update(word, f, extended, independent_left, independent_right, bow);
}

void LM::init_state(State &state){
    handler->init_state(&stagid, 1, state);
}

void LM::set_state(VocabId *ngram, unsigned char n, State &state){
    handler->init_state(ngram, n, state);
}

void LM::set_state(State &state, const State &state_prev, const Fragment *fragments, const Gap &gap){
    handler->init_state(state, state_prev, fragments, gap);
}

/* depricated */
StateId LM::get_state(VocabId *ngram, unsigned char n){
    for(unsigned char i=0;i<n;i++){
        if(ngram[i] == vocab.unk()){
            n=i;
            break;
        }
    }
    return handler->get_state(ngram, n);
}

std::size_t LM::errorcheck(std::string &pathtoarpa){
    ARPAFile *arpafile = new ARPAFile(pathtoarpa, vocab);
    logger << "[LM::errorcheck] start." << Logger::endi;
    std::size_t error_num = 0;
    size_t total = arpafile->get_totalsize();
    for(size_t i=0;i<total;i++){
        unsigned short n;
        VocabId *ngram;
        float prob;
        float bow;
        bool bow_presence = arpafile->get_ngram(n,ngram,prob,bow);
    
        bool builderror = handler->errorcheck(n, ngram, bow, prob, bow_presence);
        if(!builderror){
            logger << "[LM::build] Error: line=" << i << Logger::ende;
            error_num++;
        }
        delete [] ngram;
    }
    delete arpafile;
    logger << "[LM::build] # of errors: " << error_num << Logger::endi;
    return error_num;
}

void LM::build(std::string &pathtoarpa, std::string &pathtotreefile, unsigned int dividenum, std::string &tmpdir, unsigned int n_cores, int transposed, int prob_bit, int bow_bit) {
    logger << "[LM::build] LM build begin." << Logger::endi;

    if(v.get_opt()==DALM_OPT_REVERSE){
        handler = new ReverseDAHandler();
    }else if(v.get_opt()==DALM_OPT_EMBEDDING){
        handler = new EmbeddedDAHandler();
    }else if(v.get_opt()== DALM_OPT_BST){
        handler = new BstDAHandler();
    }else if(v.get_opt()== DALM_OPT_PT){
        handler = new PartlyTransposedDAHandler(transposed);
    }else if(v.get_opt()== DALM_OPT_QPT){

      if(prob_bit == 0) prob_bit = 32;
      if(bow_bit == 0) bow_bit = 32;

        if(prob_bit < 8){
            if(bow_bit < 9){
                handler = new QuantPartlyTransposedDAHandler<char,char>(transposed);
                logger << "[LM::build][Quant] char char " << Logger::endc;
            }else if(bow_bit < 17){
                handler = new QuantPartlyTransposedDAHandler<char,short>(transposed);
                logger << "[LM::build][Quant] char short" << Logger::endc;
            }else{
                handler = new QuantPartlyTransposedDAHandler<char,int>(transposed);
                logger << "[LM::build][Quant] char int " << Logger::endc;
            }
        }else if(prob_bit < 16){
            if(bow_bit < 9){
                handler = new QuantPartlyTransposedDAHandler<short,char>(transposed);
                logger << "[LM::build][Quant] short char " << Logger::endc;
            }else if(bow_bit < 17){
                handler = new QuantPartlyTransposedDAHandler<short,short>(transposed);
                logger << "[LM::build][Quant] short short " << Logger::endc;
            }else{
                handler = new QuantPartlyTransposedDAHandler<short,int>(transposed);
                logger << "[LM::build][Quant] short int " << Logger::endc;
            }
        }else{
            if(bow_bit < 9){
                handler = new QuantPartlyTransposedDAHandler<int,char>(transposed);
                logger << "[LM::build][Quant] int char " << Logger::endc;
            }else if(bow_bit < 17){
                handler = new QuantPartlyTransposedDAHandler<int,short>(transposed);
                logger << "[LM::build][Quant] int short " << Logger::endc;
            }else{
                handler = new QuantPartlyTransposedDAHandler<int,int>(transposed);
                logger << "[LM::build][Quant] int int " << Logger::endc;
            }
        }
    }else{
        logger << "[LM::build] DALM unknown type." << Logger::endc;
        throw "type error";
    }
    handler->build(dividenum, pathtoarpa, pathtotreefile, vocab, tmpdir, n_cores, logger);

    logger << "[LM::build] LM build end." << Logger::endi;
}

void LM::dumpParams(BinaryFileWriter &writer){
    handler->dump(writer);
}
void LM::exportParams(TextFileWriter &writer){
    handler->export_to_text(writer);
}

void LM::readParams(BinaryFileReader &reader, unsigned char order){
    if(v.get_opt() == DALM_OPT_REVERSE){
        handler = new ReverseDAHandler(reader, order, logger);
    }else if(v.get_opt() == DALM_OPT_EMBEDDING){
        handler = new EmbeddedDAHandler(reader, order, logger);
    }else if(v.get_opt() == DALM_OPT_BST){
        handler = new BstDAHandler(reader, order, logger);
    }else if(v.get_opt()== DALM_OPT_PT){
        handler = new PartlyTransposedDAHandler(reader, order, logger);
    }else if(v.get_opt()== DALM_OPT_QPT){
        reader >> DALM::QUANT_INFO::prob_bits;
        reader >> DALM::QUANT_INFO::bow_bits;

        int prob_bit = DALM::QUANT_INFO::prob_bits;
        int bow_bit = DALM::QUANT_INFO::bow_bits;

        for(int i = 0; i < DALM_MAX_ORDER-1; ++i){
            DALM::QUANT_INFO::prob_bin[i].resize(1<<prob_bit, 0);
            DALM::QUANT_INFO::bow_bin[i].resize(1<<bow_bit, 0);
            reader.read_many(DALM::QUANT_INFO::prob_bin[i].data(), 1<<prob_bit);
            reader.read_many(DALM::QUANT_INFO::bow_bin[i].data(), 1<<bow_bit);
        }


        QUANT_INFO::independent_left_mask = 1 << prob_bit;
        QUANT_INFO::prob_mask = (1<<prob_bit)-1;


        if(prob_bit < 8){
            if(bow_bit < 9){
                handler = new QuantPartlyTransposedDAHandler<char,char>(reader, order, logger);
                logger << "[LM::build][Quant] char char " << Logger::endc;
            }else if(bow_bit < 17){
                handler = new QuantPartlyTransposedDAHandler<char,short>(reader, order, logger);
                logger << "[LM::build][Quant] char short " << Logger::endc;
            }else{
                handler = new QuantPartlyTransposedDAHandler<char,int>(reader, order, logger);
                logger << "[LM::build][Quant] char int " << Logger::endc;
            }
        }else if(prob_bit < 16){
            if(bow_bit < 9){
                handler = new QuantPartlyTransposedDAHandler<short,char>(reader, order, logger);
                logger << "[LM::build][Quant] short char " << Logger::endc;
            }else if(bow_bit < 17){
                handler = new QuantPartlyTransposedDAHandler<short,short>(reader, order, logger);
                logger << "[LM::build][Quant] short short " << Logger::endc;
            }else{
                handler = new QuantPartlyTransposedDAHandler<short,int>(reader, order, logger);
                logger << "[LM::build][Quant] short int " << Logger::endc;
            }
        }else{
            if(bow_bit < 9){
                handler = new QuantPartlyTransposedDAHandler<int,char>(reader, order, logger);
                logger << "[LM::build][Quant] int char " << Logger::endc;
            }else if(bow_bit < 17){
                handler = new QuantPartlyTransposedDAHandler<int,short>(reader, order, logger);
                logger << "[LM::build][Quant] int short " << Logger::endc;
            }else{
                handler = new QuantPartlyTransposedDAHandler<int,int>(reader, order, logger);
                logger << "[LM::build][Quant] int int " << Logger::endc;
            }
        }
    }else{
        logger << "[LM::build] DALM unknown type." << Logger::endc;
        throw "type error";
    }
}

void Model::load(){
    struct stat s;
    if(stat(modeldir.c_str(), &s) != 0){
        throw std::runtime_error("Model Directory Not Found.");
    }
    std::string ini_path = modeldir + "/dalm.ini";
    std::ifstream ifs(ini_path.c_str());
    std::string line;

    std::getline(ifs, line);
    while(!ifs.eof()){
        unsigned long pos = line.find("=");
        std::string key = line.substr(0, pos);
        std::string value = line.substr(pos+1, line.size()-pos);
        if(key=="MODEL"){
            model = value;
        }else if(key=="WORDS"){
            words = value;
        }else if(key=="WORDSTXT"){
            wordstxt = value;
        }else if(key=="ORDER"){
            if(order != 0){
                assert(order == std::atoi(value.c_str()));
            }else{
                order = (unsigned char)std::atoi(value.c_str());
            }
        }
        std::getline(ifs, line);
    }

    model = modeldir + "/" + model;
    words = modeldir + "/" + words;
    wordstxt = modeldir + "/" + wordstxt;

    vocab = new Vocabulary(words, logger);
    lm = new LM(model, *vocab, order, logger);
    vocab_file = new VocabularyFile(wordstxt);
}

Model::Model(std::string modeldir, unsigned char order, Logger &logger): order(order), modeldir(modeldir), logger(logger) {
    load();
}

Model::Model(std::string modeldir, Logger &logger): modeldir(modeldir), logger(logger){
    order = 0;
    load();
}

float LM::sum_bows(State &state, unsigned char begin, unsigned char end) {
    return handler->sum_bows(state, begin, end);
}
