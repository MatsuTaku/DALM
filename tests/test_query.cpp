#include <fstream>
#include "gtest/gtest.h"
#include "../src/dalm.h"
#include "../src/dalm/buildutil.h"
#include "dalm_test_config.h"
#include "build_dalm_for_test.h"

namespace {
    void push(DALM::VocabId *ngram, size_t n, DALM::VocabId wid){
        for(size_t i = n-1; i+1 >= 1 ; i--){
            ngram[i] = ngram[i-1];
        }
        ngram[0] = wid;
    }

    class QueryTest: public ::testing::Test {
    protected:
        QueryTest(): logger(stderr), test_files(std::string(SOURCE_ROOT) + "/tests/data/test_files.txt"){
            std::ifstream ifs(test_files.c_str());
            std::getline(ifs, arpa);

            arpa = std::string(SOURCE_ROOT) + "/" + arpa;
        }

        virtual void SetUp(){
        }
        virtual void TearDown(){
        }

        DALM::Logger logger;
        std::string test_files;
        std::string arpa;
    };

    TEST_F(QueryTest, test_reverse){
        DALM::Model *m = build_dalm_for_test("rev_query", "reverse", logger, 4);

        std::string test_file = std::string(SOURCE_ROOT) + "/tests/data/kftt-tune.10.ja";
        std::string expected_file = std::string(SOURCE_ROOT) + "/tests/data/test_query_expected.txt";

        std::ifstream efile(expected_file.c_str());
        std::ifstream tfile(test_file.c_str());

        while(!efile.eof()){
            float sentence_prob;
            efile >> sentence_prob;

            std::string line;
            std::getline(tfile, line);
            std::istringstream iss(line);
            DALM::State state;
            m->lm->init_state(state);
            float prob = 0.0;
            while(!iss.eof()){
                std::string word;
                iss >> word;
                DALM::VocabId id = m->vocab->lookup(word.c_str());
                prob += m->lm->query(id, state);
            }
            DALM::VocabId s_tag_end = m->vocab->lookup("</s>");
            prob += m->lm->query(s_tag_end, state);

            EXPECT_NEAR(sentence_prob, prob, 0.001);
        }
        delete m;
    }

    TEST_F(QueryTest, test_bst){
        DALM::Model *m = build_dalm_for_test("bst_query", "bst", logger, 4);

        std::string test_file = std::string(SOURCE_ROOT) + "/tests/data/kftt-tune.10.ja";
        std::string expected_file = std::string(SOURCE_ROOT) + "/tests/data/test_query_expected.txt";

        std::ifstream efile(expected_file.c_str());
        std::ifstream tfile(test_file.c_str());

        while(!efile.eof()){
            float sentence_prob;
            efile >> sentence_prob;

            std::string line;
            std::getline(tfile, line);
            std::istringstream iss(line);
            DALM::State state;
            m->lm->init_state(state);
            float prob = 0.0;
            while(!iss.eof()){
                std::string word;
                iss >> word;
                DALM::VocabId id = m->vocab->lookup(word.c_str());
                float word_prob = m->lm->query(id, state);
                prob += word_prob;
            }
            DALM::VocabId s_tag_end = m->vocab->lookup("</s>");
            prob += m->lm->query(s_tag_end, state);

            EXPECT_NEAR(sentence_prob, prob, 0.001);
        }
        delete m;
    }

    TEST_F(QueryTest, test_reverse_single){
        DALM::Model *m = build_dalm_for_test("rev_query_single", "reverse", logger, 4);

        std::string test_file = std::string(SOURCE_ROOT) + "/tests/data/kftt-tune.10.ja";
        std::string expected_file = std::string(SOURCE_ROOT) + "/tests/data/test_query_expected.txt";

        std::ifstream efile(expected_file.c_str());
        std::ifstream tfile(test_file.c_str());

        DALM::VocabId ngram[5];
        while(!efile.eof()){
            float sentence_prob;
            efile >> sentence_prob;

            std::string line;
            std::getline(tfile, line);
            std::istringstream iss(line);
            float prob = 0.0;
            DALM::VocabId s_tag_start = m->vocab->lookup("<s>");
            std::fill(ngram, ngram + 5, s_tag_start);
            while(!iss.eof()){
                std::string word;
                iss >> word;
                DALM::VocabId id = m->vocab->lookup(word.c_str());
                push(ngram, 5, id);
                prob += m->lm->query(ngram, 5);
            }
            DALM::VocabId s_tag_end = m->vocab->lookup("</s>");
            push(ngram, 5, s_tag_end);
            prob += m->lm->query(ngram, 5);

            EXPECT_NEAR(sentence_prob, prob, 0.001);
        }
        delete m;
    }
}
