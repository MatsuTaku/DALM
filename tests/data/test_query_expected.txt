-102.479
-73.3497
-32.0694
-6.04484
-191.66
-62.7462
-48.6689
-69.77
-130.594
-70.7464